src/software/ai/hl/stp/play/ball_placement/ball_placement_play_fsm.h

```diff
@@ -0,0 +1,133 @@
+     * robots line up outside of the defense area
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     */
+    void placeBall(const Update& event);
+
+    /**
+     * Action that has the placing robot retreat after placing the ball
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     */
+    void retreat(const Update& event);
+
+    /**
+     * Guard on whether the ball is in a "dead zone"
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     *
+     * @return whether the ball is in a "dead zone"
+     */
+    bool shouldKickOffWall(const Update& event);
+
+    /**
+     * Guard on whether the placing robot has finished moving the ball into a better
+     * position with a kick
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     *
+     * @return whether the kick has been performed
+     */
+    bool kickDone(const Update& event);
+
+    /**
+     * Guard on whether the placing robot has finished placing the ball into the desired
+     * position
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     *
+     * @return whether the kick has been performed
+     */
+    bool ballPlaced(const Update& event);
+
+    /**
+     * Helper function for setting up the MoveTactics of the robots away from the ball
+     * placing robot
+     *
+     * @param event the BallPlacementPlayFSM Update event
+     */
+    void setupMoveTactics(const Update& event);
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        DEFINE_SML_STATE(StartState)
+        DEFINE_SML_STATE(KickOffWallState)
+        DEFINE_SML_STATE(PlaceBallState)
+        DEFINE_SML_STATE(RetreatState)
+        DEFINE_SML_EVENT(Update)
+
+        DEFINE_SML_ACTION(placeBall)
+        DEFINE_SML_ACTION(kickOffWall)
+        DEFINE_SML_ACTION(retreat)
+
+        DEFINE_SML_GUARD(shouldKickOffWall)
+        DEFINE_SML_GUARD(kickDone)
+        DEFINE_SML_GUARD(ballPlaced)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *StartState_S + Update_E[shouldKickOffWall_G]       = KickOffWallState_S,
+            *StartState_S + Update_E[!shouldKickOffWall_G]      = PlaceBallState_S,
+            KickOffWallState_S + Update_E[!shouldKickOffWall_G] = PlaceBallState_S,
+            KickOffWallState_S + Update_E[!kickDone_G] / kickOffWall_A =
+                KickOffWallState_S,
+            KickOffWallState_S + Update_E[kickDone_G]                = StartState_S,
+            PlaceBallState_S + Update_E[shouldKickOffWall_G]         = StartState_S,
+            PlaceBallState_S + Update_E[!ballPlaced_G] / placeBall_A = PlaceBallState_S,
+            PlaceBallState_S + Update_E[ballPlaced_G] / retreat_A    = RetreatState_S,
+            RetreatState_S + Update_E[!ballPlaced_G]                 = StartState_S,
+            RetreatState_S + Update_E[ballPlaced_G] / retreat_A      = RetreatState_S);
```

Comment:
what happens if we do
```suggestion
            RetreatState_S + Update_E[!ballPlaced_G]                 = StartState_S,
            RetreatState_S + Update_E[ballPlaced_G] / retreat_A      = X,
            X + Update_E{!ballPlaced_G] = StartState_s);
```

Reply:
Looks like it does effectively the same thing, is it better to have an explicit end state (X) ?

Reply:
generally, it's good to have an explicit end state

Reply:
I think we run stop on robots if it's in the terminate state